### PR TITLE
Synchronous loading for schema

### PR DESCRIPTION
**Pull request by kibertoad**

### CODE REVIEW

**Comment by kibertoad** [comment]
@galcohen92 @kobik Ping?..

**Comment by kobik** [comment]
> @galcohen92 @kobik Ping?..

Thanks for your PR @kibertoad!

We need to think of a way to achieve this without breaking anything.

The problem with your current implementation is that it does contain breaking change, because people are probably currently calling the init function while chaining `.then` or `.catch` afterwards. By dropping the async usage of the init function, it won't return a promise, thus, won't contain `.then` or `.catch` function.

One idea is to add a sync init while preserving the current init as it is, just like I've done here https://github.com/Zooz/express-ajv-swagger-validation/pull/50/files. My issue with this solution, is that it's kinda cheating because the dereference and parsing aren't actually taken care asynchronously, as they are just postponed to be handled later by the event loop.

2nd option, is the add sync init while maintaining the current way we're doing async init using `swagger-parser`. My issue with this way is that we're relying on too many dependencies for doing the same thing, but this is my preferred option.

A 3rd option, is wait or help on https://github.com/APIDevTools/swagger-parser/issues/54.

My preference is the 2nd option and then when `swagger-parser` would add sync call, use it instead of `json-schema-deref-sync` and `js-yaml`

What do you say?

**Comment by kibertoad** [comment]
@kobik Why can't this be handled by releasing major (explicitly breaking) version with migration guide instead? There are no reasons to maintain both versions going forward. And as it'd be semver major, it'd be an opt-in upgrade.

**Comment by kibertoad** [comment]
@kobik Any comments on the suggestion to bump major version? I imagine effort to upgrade from async to sync version of express-ajv-swagger for those who choose to do so (especially considering that it's not a dependency in need of frequent updated) is going to be considerably smaller than what needs to be done to maintain compatibility. And the benefits of sync version are quite handy.

**Comment by kibertoad** [comment]
@kobik If releasing a major version is an absolute no-go (although I would be curious to hear your opinion on that), I can add async version back on top of this PR, but ideally would prefer to keep all the tests the way there are in current PR, as creating second copy for all of them to cover async instantiation would be a nightmare.

**Comment by kobik** [comment]
@kibertoad, I think that the build should have support for both sync and async init, as there might be other use cases that users may have.

But, i've now made up my mind :)

Lets add a second async init and release a new major version of _express-ajv-swagger-validation_ that will use it.

This way we still allow both use cases with the builder, while using sync init in the swagger-validator, which makes more sense.

BTW, there is no reason to clone all the tests for both sync and async init, just a number of tests to validate it was init correctly and few validations to make sure options are also handled correctly.

Are we good with that?

**Comment by kibertoad** [comment]
@kobik Sounds great! I'm on it :D

**Comment by kibertoad** [comment]
@kobik Aaaand done!

File: README.md [hunk 1]
```diff
@@ -38,14 +39,14 @@ npm install --save api-schema-builder
...
-var apiSchemaBuilder = require('api-schema-builder');
+const apiSchemaBuilder = require('api-schema-builder');
 ```
 
 ### api-schema-builder.buildSchema(PathToSwaggerFile, options)
 
 Build schema that contains ajv validators for each endpoint, it base on swagger definition.
 
-The function return Promise.
+The function return schema object.
```
Comment (kobik) [inline README.md]: This actually would be a breaking change, which we should avoid.

File: src/index.js [hunk 2]
```diff
@@ -1,21 +1,25 @@
...
+    yaml = require('js-yaml'),
+    SwaggerParser = require('swagger-parser');
 
 const DEFAULT_SETTINGS = {
     buildRequests: true,
     buildResponses: true
 };
-function buildSchema(swaggerPath, options) {
+
+function buildSchemaAsync(swaggerPath, options) {
```
[the remark this answers is not in the document]
Comment (kibertoad) [reply src/index.js]: Could you please clarify what you meant by `Lets leave the current init as it as and add a second async init and release a new major version of express-ajv-swagger-validation that will use it.` then? Did you mean "second **sync** init"?

[the remark this answers is not in the document]
Comment (kibertoad) [reply src/index.js]: Isn't the whole idea of releasing major versions is to introduce breaking changes?

**Comment by kibertoad** [comment]
@kobik Anyway, since rename is cheap, I've changed buildSchema to be async and renamed sync one to buildSchemaSync. Does that address your concerns?

File: src/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (kobik) [reply src/index.js]: Just to clarify, we're going to release a breaking change for _express-ajv-swagger-validation_, not for the builder.

> Could you please clarify what you meant by `Lets leave the current init as it as and add a second async init and release a new major version of express-ajv-swagger-validation that will use it.` then? Did you mean "second **sync** init"?

Sorry for not being clear enough, I meant that the current init should remain unchanged, using  the original _swagger-parser_ package.

[the remark this answers is not in the document]
Comment (kibertoad) [reply src/index.js]: Sure, no problem, I already adjusted implementation not to be a breaking change. What is left now?

**Comment by kibertoad** [comment]
@kobik Appreciate it! Please let me know when new version is published, I'll submit PR for express-ajv-swagger-validator then :)

**Comment by kibertoad** [comment]
@kobik Done: https://github.com/Zooz/express-ajv-swagger-validation/pull/69

**Pull request merged**